### PR TITLE
[WPD]Regard unreachable function as a possible devirtualizable target

### PR DESCRIPTION
**Pull request by mingmingl-llvm**
https://reviews.llvm.org/D115492 skips unreachable functions and potentially allows more static de-virtualizations.  The motivation is to ignore virtual deleting destructor of abstract class (e.g., `Base::~Base()` in  https://gcc.godbolt.org/z/dWMsdT9Kz).
* Note WPD already handles most pure virtual functions (like `Base::x()` in the godbolt example above), which becomes a `__cxa_pure_virtual` in the vtable slot.

This PR proposes to undo the change, because it turns out there are other unreachable functions that a general program wants to run and fail intentionally, with `LOG(FATAL)` or `CHECK` [1] for example. While many real-world applications are encouraged to check-fail sparingly, they are allowed to do so on critical errors (e.g., misconfiguration or bug is detected during server startup).
* Implementation-wise, this PR keeps the one-bit 'unreachable' state in bitcode and updates WPD analysis.
 
https://gcc.godbolt.org/z/T1aMhczYr is a minimum reproducible example extracted from unit test. `Base::func` is a one-liner of `LOG(FATAL) << "message"`, and lowered to one basic block ending with `unreachable`. A real-world program is _allowed_ to invoke Base::func to terminate the program as a way to report errors (in server initialization stage for example), even if errors on the serving path should be handled more gracefully.

[1] https://abseil.io/docs/cpp/guides/logging#CHECK and https://abseil.io/docs/cpp/guides/logging#configuration-and-flags

### CODE REVIEW

File: llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp [hunk 1]
```diff
@@ -167,6 +167,18 @@ static cl::list<std::string>
                       cl::desc("Prevent function(s) from being devirtualized"),
                       cl::Hidden, cl::CommaSeparated);
 
+/// A function is unreachable if its entry block ends with 'unreachable' IR
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp:170]: Can you note that this is useful for removing abstract class virtual destructors, which are emitted with a trap and unreachable instruction, and add a TODO to identify these more precisely?

Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp:170]: related - probably update the description which currently says that this is about pure virtual functions, but most of those are already handled due to the __cxa_pure_virtual in the vtable slot.

**Review by teresajohnson — APPROVED** [review]
lgtm with a couple of comment suggestions

File: llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp (hunk 1 above)
Comment (minglotus-6) [reply llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp]: Updated the surrounding comment with the motivating case and a TODO for Clang's codegen, and updated PR description as suggested.

**Review by mingmingl-llvm — COMMENTED** [review]
thanks for looking into this together and figuring out a viable solution forward!

**Comment by minglotus-6** [comment]
I didn't find a test failure in the buildkite log (with `cat <buildkite>.log | grep -B 10 -A 10  "Failed Tests "`).

Local tests pass. Will merge.

**Pull request merged**